File: test_braces.py
with open('src/App.tsx', 'r') as f:
    code = f.read()

stack = []
for i, c in enumerate(code):
    line = code[:i].count('\n') + 1
    if c in '{([':
        stack.append((c, line))
    elif c in '})]':
        if not stack:
            print(f"Extra {c} at line {line}")
            continue
        last, l_line = stack.pop()
        if (c == '}' and last != '{') or (c == ')' and last != '(') or (c == ']' and last != '['):
            print(f"Mismatch {c} at line {line} (last opened {last} at line {l_line})")

print(f"Left in stack: {len(stack)}")
for item in stack:
    print(item)

File: test_braces.js
const fs = require('fs');
const code = fs.readFileSync('src/App.tsx', 'utf8');
let stack = [];
for(let i=0; i<code.length; i++) {
  if (code[i] === '{') stack.push({char: '{', line: code.substring(0, i).split('\n').length});
  if (code[i] === '(') stack.push({char: '(', line: code.substring(0, i).split('\n').length});
  if (code[i] === '[') stack.push({char: '[', line: code.substring(0, i).split('\n').length});
  
  if (code[i] === '}') {
    if(stack[stack.length-1].char === '{') stack.pop();
    else console.log('Mismatch } at line ' + code.substring(0, i).split('\n').length + ' expected ' + stack[stack.length-1].char);
  }
  if (code[i] === ')') {
    if(stack[stack.length-1].char === '(') stack.pop();
    else console.log('Mismatch ) at line ' + code.substring(0, i).split('\n').length + ' expected ' + stack[stack.length-1].char);
  }
  if (code[i] === ']') {
    if(stack[stack.length-1].char === '[') stack.pop();
    else console.log('Mismatch ] at line ' + code.substring(0, i).split('\n').length + ' expected ' + stack[stack.length-1].char);
  }
}
console.log('Left in stack:', stack.length);
